File: Sources/SnipKeyCore/SnippetTriggerSuggester.swift
import Foundation

public enum SnippetTriggerSuggester {
    public static let defaultPreferredLength = 4

    private static let defaultBase = "clip"
    private static let genericHostLabels: Set<String> = ["www", "m", "mobile", "app"]
    private static let genericSecondLevelDomains: Set<String> = ["ac", "co", "com", "edu", "gov", "net", "org"]
    private static let emailRegex = try? NSRegularExpression(
        pattern: "^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,}$",
        options: [.caseInsensitive]
    )
    private static let linkDetector = try? NSDataDetector(types: NSTextCheckingResult.CheckingType.link.rawValue)

    public static func suggestTrigger(
        for content: String,
        existingTriggers: [String] = [],
        preferredLength: Int = defaultPreferredLength
    ) -> String {
        let resolvedPreferredLength = max(1, preferredLength)
        let normalizedExisting = Set(existingTriggers.map { $0.lowercased() })

        for candidate in candidates(for: content) {
            let sanitized = sanitize(candidate.rawValue)
            guard sanitized.isEmpty == false else { continue }
            return makeUnique(
                base: sanitized,
                direction: candidate.direction,
                existingTriggers: normalizedExisting,
                preferredLength: resolvedPreferredLength
            )
        }

        return makeUnique(
            base: defaultBase,
            direction: .prefix,
            existingTriggers: normalizedExisting,
            preferredLength: resolvedPreferredLength
        )
    }

    private static func candidates(for content: String) -> [TriggerCandidate] {
        let collapsed = collapseWhitespace(in: content)
        guard collapsed.isEmpty == false else {
            return [TriggerCandidate(rawValue: defaultBase, direction: .prefix)]
        }

        var candidates: [TriggerCandidate] = []

        if let code = codeCandidate(from: collapsed) {
            candidates.append(TriggerCandidate(rawValue: code, direction: .suffix))
        }

        if let email = emailCandidate(from: collapsed) {
            candidates.append(TriggerCandidate(rawValue: email, direction: .prefix))
        }

        candidates.append(contentsOf: urlCandidates(from: collapsed).map {
            TriggerCandidate(rawValue: $0, direction: .prefix)
        })

        if let filePath = pathCandidate(from: collapsed) {
            candidates.append(TriggerCandidate(rawValue: filePath, direction: .prefix))
        }

        if let englishPhrase = englishPhraseCandidate(from: collapsed) {
            candidates.append(TriggerCandidate(rawValue: englishPhrase, direction: .prefix))
        }

        if containsCJK(in: collapsed), let initials = cjkInitialsCandidate(from: collapsed) {
            candidates.append(TriggerCandidate(rawValue: initials, direction: .prefix))
        }

        if let identifier = identifierCandidate(from: collapsed) {
            candidates.append(TriggerCandidate(rawValue: identifier, direction: .prefix))
        }

        candidates.append(TriggerCandidate(rawValue: defaultBase, direction: .prefix))
        return deduplicated(candidates)
    }

    private static func codeCandidate(from content: String) -> String? {
        guard content.isEmpty == false else { return nil }
        guard content.unicodeScalars.contains(where: { $0.isASCII && CharacterSet.alphanumerics.contains($0) }) else { return nil }
        guard content.unicodeScalars.allSatisfy({ scalar in
            scalar.isASCII && CharacterSet.alphanumerics.contains(scalar)
        }) else {
            return nil
        }

        let letters = content.filter(\.isLetter).count
        let digits = content.filter(\.isNumber).count
        let hasUppercaseLetter = content.contains { $0.isUppercase }

        guard content.count >= 6 else { return nil }
        guard digits >= 4 else { return nil }
        guard hasUppercaseLetter || letters == 0 || digits >= letters * 2 else { return nil }

        return content.lowercased()
    }

    private static func emailCandidate(from content: String) -> String? {
        guard isLikelyEmail(content) else { return nil }
        let localPart = content.split(separator: "@", maxSplits: 1).first.map(String.init) ?? ""
        return identifierCandidate(from: localPart)
    }

    private static func isLikelyEmail(_ content: String) -> Bool {
        guard let emailRegex else {
            return false
        }

        let range = NSRange(content.startIndex..., in: content)
        return emailRegex.firstMatch(in: content, options: [], range: range)?.range == range
    }

    private static func urlCandidates(from content: String) -> [String] {
        guard let linkDetector else {
            return []
        }

        let fullRange = NSRange(content.startIndex..., in: content)
        guard let match = linkDetector.firstMatch(in: content, options: [], range: fullRange),
              match.range == fullRange,
              let url = match.url,
              url.scheme?.lowercased() != "mailto",
              let host = url.host else {
            return []
        }

        let hostTokens = host
            .lowercased()
            .split(separator: ".")
            .map(String.init)
            .filter { genericHostLabels.contains($0) == false }

        guard let domain = preferredDomainLabel(from: hostTokens) else { return [] }

        var candidates = [domain]
        if let pathToken = preferredURLPathToken(from: url.path) {
            candidates.insert(domain + pathToken, at: 0)
        }

        return candidates.compactMap(identifierCandidate(from:))
    }

    private static func preferredDomainLabel(from hostTokens: [String]) -> String? {
        guard hostTokens.isEmpty == false else { return nil }
        if hostTokens.count == 1 { return hostTokens[0] }

        let labelsBeforeTLD = hostTokens.dropLast().reversed()
        for label in labelsBeforeTLD where genericSecondLevelDomains.contains(label) == false {
            return label
        }

        return hostTokens.dropLast().last
    }

    private static func preferredURLPathToken(from path: String) -> String? {
        let components = path
            .split(separator: "/")
            .map(String.init)
            .filter { $0.isEmpty == false }

        for component in components.reversed() {
            if let candidate = identifierCandidate(from: component), candidate.count >= 3 {
                return candidate
            }
        }

        return nil
    }

    private static func pathCandidate(from content: String) -> String? {
        guard looksLikePath(content) else { return nil }

        let resolvedPath: String
        if content.lowercased().hasPrefix("file://"), let url = URL(string: content), url.isFileURL {
            resolvedPath = url.path
        } else {
            resolvedPath = (content as NSString).expandingTildeInPath
        }

        let lastPathComponent = (resolvedPath as NSString).lastPathComponent
        let stem = (lastPathComponent as NSString).deletingPathExtension
        return identifierCandidate(from: stem)
    }

    private static func looksLikePath(_ content: String) -> Bool {
        if content.hasPrefix("/") || content.hasPrefix("~/") || content.lowercased().hasPrefix("file://") {
            return true
        }

        return content.contains("/") && content.contains(where: { $0.isWhitespace }) == false && content.contains("://") == false
    }

    private static func englishPhraseCandidate(from content: String) -> String? {
        guard containsCJK(in: content) == false else { return nil }
        guard content.contains(where: { $0.isWhitespace }) else { return nil }

        let tokens = wordTokens(from: transliteratedASCII(from: content)).filter { token in
            token.isEmpty == false && token.allSatisfy({ $0.isNumber }) == false
        }

        guard tokens.count > 1 else { return nil }

        let initials = tokens.compactMap(\.first)
        guard initials.isEmpty == false else { return nil }
        return String(initials)
    }

    private static func identifierCandidate(from content: String) -> String? {
        let normalized = transliteratedASCII(from: content).filter { $0.isLetter || $0.isNumber }
        return normalized.isEmpty ? nil : normalized
    }

    private static func cjkInitialsCandidate(from content: String) -> String? {
        var initials = ""

        for character in content {
            let transliterated = transliteratedASCII(from: String(character))
            guard let initial = transliterated.first(where: { $0.isLetter || $0.isNumber }) else {
                continue
            }

            initials.append(initial)
        }

        return initials.isEmpty ? nil : initials
    }

    private static func collapseWhitespace(in content: String) -> String {
        content
            .replacingOccurrences(of: "\n", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .components(separatedBy: .whitespacesAndNewlines)
            .filter { $0.isEmpty == false }
            .joined(separator: " ")
    }

    private static func transliteratedASCII(from content: String) -> String {
        let mutable = NSMutableString(string: content) as CFMutableString
        CFStringTransform(mutable, nil, kCFStringTransformToLatin, false)
        CFStringTransform(mutable, nil, kCFStringTransformStripCombiningMarks, false)
        return (mutable as String).lowercased()
    }

    private static func wordTokens(from content: String) -> [String] {
        content.split { character in
            character.isLetter == false && character.isNumber == false
        }.map(String.init)
    }

    private static func sanitize(_ candidate: String) -> String {
        let transliterated = transliteratedASCII(from: candidate)
        let result = transliterated.filter { $0.isLetter || $0.isNumber }
        return result
    }

    private static func makeUnique(
        base: String,
        direction: TriggerDirection,
        existingTriggers: Set<String>,
        preferredLength: Int
    ) -> String {
        let resolvedLength = min(max(1, preferredLength), base.count)

        for length in resolvedLength...base.count {
            let candidate = candidateSlice(from: base, direction: direction, length: length)
            if existingTriggers.contains(candidate.lowercased()) == false {
                return candidate
            }
        }

        return makeNumericUnique(
            from: candidateSlice(from: base, direction: direction, length: resolvedLength),
            existingTriggers: existingTriggers
        )
    }

    private static func candidateSlice(from base: String, direction: TriggerDirection, length: Int) -> String {
        switch direction {
        case .prefix:
            return String(base.prefix(length))
        case .suffix:
            return String(base.suffix(length))
        }
    }

    private static func makeNumericUnique(from base: String, existingTriggers: Set<String>) -> String {
        var index = 2
        while true {
            let candidate = base + String(index)
            if existingTriggers.contains(candidate.lowercased()) == false {
                return candidate
            }
            index += 1
        }
    }

    private static func deduplicated(_ candidates: [TriggerCandidate]) -> [TriggerCandidate] {
        var seen = Set<String>()
        var result: [TriggerCandidate] = []

        for candidate in candidates {
            let key = candidate.direction.rawValue + ":" + candidate.rawValue.lowercased()
            if seen.insert(key).inserted {
                result.append(candidate)
            }
        }

        return result
    }

    private static func containsCJK(in content: String) -> Bool {
        content.unicodeScalars.contains { scalar in
            switch scalar.value {
            case 0x3400...0x4DBF, 0x4E00...0x9FFF, 0xF900...0xFAFF:
                return true
            default:
                return false
            }
        }
    }

    private struct TriggerCandidate {
        let rawValue: String
        let direction: TriggerDirection
    }

    private enum TriggerDirection: String {
        case prefix
        case suffix
    }
}

File: Windows/SnipKey.Windows/Core/SnippetTriggerSuggester.cs
using System.Globalization;
using System.IO;
using System.Text;
using System.Text.RegularExpressions;

namespace SnipKey.WinApp.Core;

public static partial class SnippetTriggerSuggester
{
    public const int DefaultPreferredLength = 4;

    private const string DefaultBase = "clip";
    private static readonly HashSet<string> GenericHostLabels = ["www", "m", "mobile", "app"];
    private static readonly HashSet<string> GenericSecondLevelDomains = ["ac", "co", "com", "edu", "gov", "net", "org"];

    public static string SuggestTrigger(string content, IEnumerable<string>? existingTriggers = null, int preferredLength = DefaultPreferredLength)
    {
        var resolvedPreferredLength = Math.Max(1, preferredLength);
        var normalizedExisting = (existingTriggers ?? [])
            .Select(trigger => trigger.ToLowerInvariant())
            .ToHashSet();

        foreach (var candidate in Candidates(content))
        {
            var sanitized = Sanitize(candidate.RawValue);
            if (string.IsNullOrEmpty(sanitized))
            {
                continue;
            }

            return MakeUnique(sanitized, candidate.Direction, normalizedExisting, resolvedPreferredLength);
        }

        return MakeUnique(DefaultBase, TriggerDirection.Prefix, normalizedExisting, resolvedPreferredLength);
    }

    private static List<TriggerCandidate> Candidates(string content)
    {
        var collapsed = CollapseWhitespace(content);
        if (string.IsNullOrEmpty(collapsed))
        {
            return [new TriggerCandidate(DefaultBase, TriggerDirection.Prefix)];
        }

        var candidates = new List<TriggerCandidate>();

        if (CodeCandidate(collapsed) is { } code)
        {
            candidates.Add(new TriggerCandidate(code, TriggerDirection.Suffix));
        }

        if (EmailCandidate(collapsed) is { } email)
        {
            candidates.Add(new TriggerCandidate(email, TriggerDirection.Prefix));
        }

        candidates.AddRange(UrlCandidates(collapsed).Select(candidate => new TriggerCandidate(candidate, TriggerDirection.Prefix)));

        if (PathCandidate(collapsed) is { } path)
        {
            candidates.Add(new TriggerCandidate(path, TriggerDirection.Prefix));
        }

        if (EnglishPhraseCandidate(collapsed) is { } englishPhrase)
        {
            candidates.Add(new TriggerCandidate(englishPhrase, TriggerDirection.Prefix));
        }

        if (IdentifierCandidate(collapsed) is { } identifier)
        {
            candidates.Add(new TriggerCandidate(identifier, TriggerDirection.Prefix));
        }

        candidates.Add(new TriggerCandidate(DefaultBase, TriggerDirection.Prefix));
        return Deduplicated(candidates);
    }

    private static string? CodeCandidate(string content)
    {
        if (content.Length < 6 || content.Any(character => character > 127 || !char.IsLetterOrDigit(character)))
        {
            return null;
        }

        var letters = content.Count(char.IsLetter);
        var digits = content.Count(char.IsDigit);
        var hasUppercaseLetter = content.Any(char.IsUpper);

        if (digits < 4 || (!hasUppercaseLetter && letters != 0 && digits < letters * 2))
        {
            return null;
        }

        return content.ToLowerInvariant();
    }

    private static string? EmailCandidate(string content)
    {
        if (!EmailRegex().IsMatch(content))
        {
            return null;
        }

        var localPart = content.Split('@', 2)[0];
        return IdentifierCandidate(localPart);
    }

    private static IEnumerable<string> UrlCandidates(string content)
    {
        if (!Uri.TryCreate(content, UriKind.Absolute, out var uri) || uri.Scheme.Equals("mailto", StringComparison.OrdinalIgnoreCase))
        {
            return [];
        }

        if (string.IsNullOrEmpty(uri.Host))
        {
            return [];
        }

        var hostTokens = uri.Host
            .ToLowerInvariant()
            .Split('.', StringSplitOptions.RemoveEmptyEntries)
            .Where(token => !GenericHostLabels.Contains(token))
            .ToList();
        var domain = PreferredDomainLabel(hostTokens);
        if (domain is null)
        {
            return [];
        }

        var candidates = new List<string> { domain };
        if (PreferredUrlPathToken(uri.AbsolutePath) is { } pathToken)
        {
            candidates.Insert(0, domain + pathToken);
        }

        return candidates.Select(IdentifierCandidate).Where(candidate => candidate is not null).Cast<string>();
    }

    private static string? PreferredDomainLabel(IReadOnlyList<string> hostTokens)
    {
        if (hostTokens.Count == 0)
        {
            return null;
        }

        if (hostTokens.Count == 1)
        {
            return hostTokens[0];
        }

        foreach (var label in hostTokens.Take(hostTokens.Count - 1).Reverse())
        {
            if (!GenericSecondLevelDomains.Contains(label))
            {
                return label;
            }
        }

        return hostTokens[^2];
    }

    private static string? PreferredUrlPathToken(string path)
    {
        var components = path.Split('/', StringSplitOptions.RemoveEmptyEntries);
        foreach (var component in components.Reverse())
        {
            var candidate = IdentifierCandidate(component);
            if (candidate is { Length: >= 3 })
            {
                return candidate;
            }
        }

        return null;
    }

    private static string? PathCandidate(string content)
    {
        if (!LooksLikePath(content))
        {
            return null;
        }

        var resolvedPath = content.StartsWith("file://", StringComparison.OrdinalIgnoreCase) && Uri.TryCreate(content, UriKind.Absolute, out var uri)
            ? uri.LocalPath
            : content;
        var fileName = Path.GetFileNameWithoutExtension(resolvedPath.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar));
        return IdentifierCandidate(fileName);
    }

    private static bool LooksLikePath(string content)
    {
        if (content.StartsWith("/", StringComparison.Ordinal)
            || content.StartsWith("~/", StringComparison.Ordinal)
            || content.StartsWith("file://", StringComparison.OrdinalIgnoreCase)
            || Regex.IsMatch(content, "^[A-Za-z]:[\\\\/]"))
        {
            return true;
        }

        return content.Contains('/') && !content.Any(char.IsWhiteSpace) && !content.Contains("://", StringComparison.Ordinal);
    }

    private static string? EnglishPhraseCandidate(string content)
    {
        if (ContainsCjk(content) || !content.Any(char.IsWhiteSpace))
        {
            return null;
        }

        var tokens = WordTokens(TransliteratedAscii(content))
            .Where(token => token.Length > 0 && !token.All(char.IsDigit))
            .ToList();
        if (tokens.Count <= 1)
        {
            return null;
        }

        return new string(tokens.Select(token => token[0]).ToArray());
    }

    private static string? IdentifierCandidate(string content)
    {
        var normalized = new string(TransliteratedAscii(content).Where(char.IsLetterOrDigit).ToArray());
        return string.IsNullOrEmpty(normalized) ? null : normalized;
    }

    private static string CollapseWhitespace(string content)
    {
        return string.Join(' ', content.Replace("\n", " ", StringComparison.Ordinal)
            .Trim()
            .Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries));
    }

    private static string TransliteratedAscii(string content)
    {
        var normalized = content.Normalize(NormalizationForm.FormD);
        var builder = new StringBuilder(normalized.Length);
        foreach (var character in normalized)
        {
            var category = CharUnicodeInfo.GetUnicodeCategory(character);
            if (category == UnicodeCategory.NonSpacingMark)
            {
                continue;
            }

            if (character <= 127)
            {
                builder.Append(char.ToLowerInvariant(character));
            }
        }

        return builder.ToString();
    }

    private static IEnumerable<string> WordTokens(string content)
    {
        return Regex.Split(content, "[^A-Za-z0-9]+").Where(token => token.Length > 0);
    }

    private static string Sanitize(string candidate)
    {
        return new string(TransliteratedAscii(candidate).Where(char.IsLetterOrDigit).ToArray());
    }

    private static string MakeUnique(string baseValue, TriggerDirection direction, HashSet<string> existingTriggers, int preferredLength)
    {
        var resolvedLength = Math.Min(Math.Max(1, preferredLength), baseValue.Length);
        for (var length = resolvedLength; length <= baseValue.Length; length += 1)
        {
            var candidate = CandidateSlice(baseValue, direction, length);
            if (!existingTriggers.Contains(candidate.ToLowerInvariant()))
            {
                return candidate;
            }
        }

        return MakeNumericUnique(CandidateSlice(baseValue, direction, resolvedLength), existingTriggers);
    }

    private static string CandidateSlice(string baseValue, TriggerDirection direction, int length)
    {
        return direction == TriggerDirection.Prefix ? baseValue[..length] : baseValue[^length..];
    }

    private static string MakeNumericUnique(string baseValue, HashSet<string> existingTriggers)
    {
        var index = 2;
        while (true)
        {
            var candidate = baseValue + index.ToString(CultureInfo.InvariantCulture);
            if (!existingTriggers.Contains(candidate.ToLowerInvariant()))
            {
                return candidate;
            }

            index += 1;
        }
    }

    private static List<TriggerCandidate> Deduplicated(IEnumerable<TriggerCandidate> candidates)
    {
        var seen = new HashSet<string>();
        var result = new List<TriggerCandidate>();
        foreach (var candidate in candidates)
        {
            var key = candidate.Direction + ":" + candidate.RawValue.ToLowerInvariant();
            if (seen.Add(key))
            {
                result.Add(candidate);
            }
        }

        return result;
    }

    private static bool ContainsCjk(string content)
    {
        return content.Any(character => character is >= '\u3400' and <= '\u4DBF' or >= '\u4E00' and <= '\u9FFF' or >= '\uF900' and <= '\uFAFF');
    }

    [GeneratedRegex("^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,}$", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant)]
    private static partial Regex EmailRegex();

    private readonly record struct TriggerCandidate(string RawValue, TriggerDirection Direction);

    private enum TriggerDirection
    {
        Prefix,
        Suffix
    }
}